### PR TITLE
chore: generate CI config & use context dir

### PR DESCRIPTION
**Pull request by shykes**
This is a small first step towards eventually generating 100% of our Github Actions config.

- Docs
- SDK checks: test, lint, test-publish (on stable and dev engine)

### CODE REVIEW

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (jedevc) [inline .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: We could add all these files into a `.gitattributes` in this folder, that way changes to them won't show up as massive diffs, etc :tada:

Comment (jedevc) [inline .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: Would it possible to *group* multiple dagger commands into a single file? e.g. I think it would be really nice to have all the `elixir` checks in one single pipeline, but potentially all different jobs, or maybe even steps in a single job?

Not 100% sure, I don't think we need to be optimizing to have a pretty view in github. But mainly, I think it would help avoid spinning up too many new CI workers (which can be pricey).

File: .github/main.go
Comment (jedevc) [inline .github/main.go:52]: This is *so* neat. I really really really like this.

File: .github/main.go
Comment (jedevc) [inline .github/main.go:42]: Potential interface idea - instead of needing to have a bunch of booleans here, we could maybe have a custom enum of event types in the `gha` module? Then we just pass a list of `Events []GhaEvent` in (which also lets you more easily programmatically control these).

File: .github/main.go
Comment (jedevc) [inline .github/main.go:167]: A little ask - could we also use the `dirdiff` module to implement a `lint` action here as well (and *maybe* have a pre-defined workflow to make sure this in sync?).

I wish there was a way to avoid needing to have these generated files in the repo at all, I might have some fun ideas for that (hackathon project?), but they involve way too much hackery at the github actions level for now.

**Review by jedevc — COMMENTED** [review]
My only major complaint with this PR is that I didn't get to do it :laughing: Seriously, it's such a neat approach, I'm a fan :folding_hand_fan: 

This isn't a real review, just was reading through and left some thoughts about some future features/fixes - none of them are *really* blockers. The only more major point is the one around the fact that I *think* this increases our CI cost by spinning up more workers (but I left a suggestion as to what we might do about that).

I think we should try and get this in... relatively soon? Maybe sometime next week works? I think starting with just the `docs` and `sdk` steps is good, since those are "simpler", but also, means that if we turn out to not like the approach (very doubtful in my mind), we can more easily backtrack.

File: .github/workflows/docs.yml
Comment (jedevc) [inline .github/workflows/docs.yml:83]: Color? :pleading_face:

File: .github/main.go
Comment (shykes) [reply .github/main.go:167]: I actually implemented this already 😁 kudos to @kpenfound for helping figure it out. Unfortunately it's not that great, the best gha can do is generate the files and commit/push them for you. Unlike other CIs it doesn't support continuation.

Comment (jedevc) [reply .github/main.go:167]: Ahhhh, yeah that's a neat approach. That's annoying it's not great, I think we can just stick with manual invocation for now :cry: (definitely not worth blocking on)

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (shykes) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: Good idea. I guess we need to match the generated files on something, any suggestion for the pattern?

Comment (shykes) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: It's possible but requires some extra implementation work, to "collapse" pipelines that can be collapsed. Either automatically or with a "hint" from the caller. I would love a PR on the gha module with that feature :)

File: .github/main.go
Comment (shykes) [reply .github/main.go:42]: Yes that's a great idea. My only concern is: what do we do about the `onXYZ` arguments that are not boolean? Like `onPullRequestBranch` etc. How do we avoid making the split feel weird to callers?

File: .github/main.go
Comment (shykes) [reply .github/main.go:167]: I think this gap is a possible insertion point for bridging to pocketCI :) But I don't know exactly how. cc @kpenfound @matipan

File: .github/workflows/docs.yml
Comment (shykes) [reply .github/workflows/docs.yml:83]: I needed to disable color to avoid polluting my markdown report - did you see that feature?

I realize now it's a tradeoff between 1) neat logs in markdown report, and 2) colors in GHA logs

File: .github/main.go
Comment (shykes) [reply .github/main.go]: @jedevc I added a `Check()` function that uses `dirdiff`. But I wasn't sure where to integrate it in the main module, so I didn't wire it in. I could use your guidance on that.

**Review by gerhard — APPROVED** [review]
I like the idea of using Dagger to generate GitHub Actions YAML. I am not excited about removing `180` lines of YAML, and adding `11k` lines of generated YAML.

Some of the new scripting is significantly more complex than what we had before. While `action.yml` would not win any scripting competitions, it was simpler.

The only real blocker that I see to merging this is losing the `timeout` setting. Restricting `GITHUB_TOKEN` permissions would be nice to keep, but less of a blocker.

Overall, I am curious to see how well this works in practice. Specifically:
- When a job fails, is it easier to understand which step failed & why?
- How well does the workflow dispatch work after the test & lint split?
- What impact does this split have on scheduling jobs? We will need 2x as many runners.
- What's it like to live with 2 modules, `.dagger` and `.github` ?

image

One thing which I know will hurt is when we hit GitHub's `100` checks limit. After this change, we are at `91`. GitHub's UI makes it impossible to make sense of more than 100 checks. I don't think the number goes higher than 100 (this limitation goes back to 2021), and the small scrolling area makes it very difficult to find anything. Also, the `Checks` tab is getting more difficult to use:

image

I am going to approve this optimistically so that we can merge this (after resolving the conflicts) and try living with this change for a few weeks. YOLO!

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (shykes) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: @jedevc thinking about this more:

> I think it would be really nice to have all the elixir checks in one single pipeline, but potentially all different jobs

This is possible (with some implementation work) BUT it will NOT change the number of machines started, because each job gets its own machine, regardless of workflow grouping.

> or maybe even steps in a single job?

There is no benefit to this. It should be done at the Dagger level.

File: .github/main.go
Comment (shykes) [reply .github/main.go:1]: Good catch. I actually implemented `timeout` already, I just forgot to use it in the code :) Will fix now

File: .github/workflows/docs.yml
Comment (shykes) [reply .github/workflows/docs.yml:16]: Makes sense. Will do.

File: .github/main.go
Comment (shykes) [reply .github/main.go:174]: I considered that too, but that implies a quantitative increase (like memory, cpus) and some of the dimensions are more subtle - multi-tenant/single-tenant, spot/nospot etc. This felt safer and more flexible for you.

File: .github/main.go
Comment (shykes) [reply .github/main.go:112]: It's because "spot instance" became a generic term for a well-known pattern beyond AWS (instances that can be preempted depending on demand and the price you set). Whereas "on demand" is AWS-specific jargon that doesn't carry the same obvious meaning industry-wide.

So `noSpot` will be more obvious to more people.

**Comment by shykes** [comment]
Thanks Gerhard! A few notes below.

> I like the idea of using Dagger to generate GitHub Actions YAML. I am not excited about removing `180` lines of YAML, and adding `11k` lines of generated YAML.

A more fair comparison is that I replaced 180 lines of YAML with 196 lines of Go, which can collapse to a `13` line overview:

```golang
ci.Gha = dag.Gha(dagger.GhaOpts{
		DaggerVersion: daggerVersion,
		PublicToken:   "dag_dagger_sBIv6DsjNerWvTqt2bSFeigBUqWxp9bhh3ONSSgeFnw",
		Runner:        ci.BronzeRunner(false),
	})
	return ci.
		WithPipeline("Docs", "docs lint", "", false).
		WithSdkPipelines("python").
		WithSdkPipelines("typescript").
		WithSdkPipelines("go").
		WithSdkPipelines("java").
		WithSdkPipelines("elixir").
		WithSdkPipelines("rust").
		WithSdkPipelines("php")
```

I expect the ratio will only get better as switch more pipelines to be generated.

This would be nicer if Github Actions supported continuations - then the generation step could be hidden better. We'll start with proper `.gitattributes`. But certainly, to appreciate this change, you need to adopt a mindset where the generated YAML is not source code: it's like a compiled binary.

> Some of the new scripting is significantly more complex than what we had before. While `action.yml` would not win any scripting competitions, it was simpler.

It was not actually simpler. The perceived increase in complexity comes from several things:

1. Additional features. In particular the markdown report requires a larger shell script. The `action.yaml` doesn't have that feature, so it looks simpler.
2. Inlining. Most of the generated shell script is actually *identical* to the manually edited one, or only slightly tweaked. But it's all shown in one place, instead of spread out across an action YAML, a workflow YAML, and in some cases one or more intermediary workflow YAML.
3. Simpler YAML. I transferred some of the complexity from proprietary YAML to Bash. This makes the bash script look slightly more complicated (not that much really). But in return we get less context switching between YAML and bash, and more local debugability. For example you can debug the markdown generation feature with `COMMAND='echo hello world' ./scripts/exec.sh; less github-summary.md`. Same for all the other steps.

Some of the above ☝️ are not specific to generating the YAML. For example, I *could* generate a workflow that references a shared static `action.yaml`, or intermediary workflow. I just chose not to, because I like this tradeoff. Operationally one single self-contained YAML, with everything inlined and no fancy runtime abstractions, is ultimately more reliable. But, even if that turns out to be wrong, we can keep the generator model and just change the layout that we generate to. We could even support several layouts, and make it configurable.

> The only real blocker that I see to merging this is losing the `timeout` setting. Restricting `GITHUB_TOKEN` permissions would be nice to keep, but less of a blocker.

Thanks for catching that. I'm on it.

> Overall, I am curious to see how well this works in practice. Specifically:
> 
> * When a job fails, is it easier to understand which step failed & why?

So far it seems the be either the same or better, depending on the situation.

- The SDK split is nice
- More standardized layout across workflows is a subtle advantage. I know what to expect across all checks, it's the same structure, so less cognitive load when debugging
- The markdown summary is very useful, I often never go to the logs at all, I go straight to the summary (sadly github UI makes that 2 clicks from the PR)

> * How well does the workflow dispatch work after the test & lint split?

Not sure what you mean by that.

> * What impact does this split have on scheduling jobs? We will need 2x as many runners.

I don't think so. My understanding is that machines are allocated at the job level, not the workflow level. This is one reason I went for one pipeline per workflow file. It didn't seem to impact the runtime behavior that much.

> * What's it like to live with 2 modules, `.dagger` and `.github` ?

I am not attached to 2 modules particularly, but it seems most aligned with the current consensus on how to tackle code generation pipelines with Dagger. That is totally something we can change.

> One thing which I know will hurt is when we hit GitHub's `100` checks limit. After this change, we are at `91`. GitHub's UI makes it impossible to make sense of more than 100 checks. I don't think the number goes higher than 100 (this limitation goes back to 2021), and the small scrolling area makes it very difficult to find anything. Also, the `Checks` tab is getting more difficult to use:

I actually think we can bring the number of checks down easily. I didn't mind the increased number of checks (if anything I started to look at the Dagger Cloud checks less, because I get the same granularity from GHA checks now). But I can prioritize bringing the number the checks down.

The easiest way to do this, is to collapse pipelines in Dagger. Lint, test and test-publish for each SDK can be collapsed into a single dagger call. Splitting that across 3 VMs per SDK is overkill anyway.

Separately, we should talk about what this model means for Dagger Cloud checks. Do we still need that granularity in a world where generated CI config is the norm, giving you the granularity for free? Maybe we can switch to a single Dagger Cloud check that links to the entire trace group? cc @vito @aluzzardi @kpenfound @crjm 

> I am going to approve this optimistically so that we can merge this (after resolving the conflicts) and try living with this change for a few weeks. YOLO!

Thank you, I appreciate the detailed review!

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (shykes) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: Done

File: .github/main.go
Comment (shykes) [reply .github/main.go:1]: Fixed!

File: .github/workflows/docs.yml
Comment (shykes) [reply .github/workflows/docs.yml:16]: @gerhard Done :) https://github.com/dagger/dagger/pull/8241/commits/c78537f1404c6b6f3460f29977c2caad95e22f4e

File: .github/main.go
Comment (jedevc) [reply .github/main.go]: Hm, I'd add a new github workflow for right now for this - something like `github lint`? I suppose wiring it into the main module makes sense, but we could also just have a single workflow that directly calls the github module, I'm not hugely biased towards either.

File: .github/main.go
Comment (jedevc) [reply .github/main.go:61]: So what I can tell - the event block here is made up of two things - the thing that can happen to the branch, and the branch target.

So *actually* it's a bit trickier than an enum - it's almost like we want an event to be:

```go
type GithubEvent struct {
	PullRequest *GithubPullRequestEvent
	Issue *IssueEvent
	... // more events
}

type GithubPullRequestEvent struct {
	Branches []string
	Triggers []GithubPullRequestTrigger
}

type GithubPullRequestTrigger string
const (
	GithubPullRequestOpened = "opened"
	// ...
)
```

Then, you just pass in a list of `GithubEvent` to your `WithPipeline` - maybe constructing them something like:

```
event := dag.OnPullRequest().WithBranch("main").OnOpen().OnClose()
```

How does something like that sound? Feels very literate-style, but then each event is it's own dagger object to manipulate (which would be *really* cool, now you could build little functions to encapsulate the logic for your project).

File: .github/workflows/docs.yml
Comment (jedevc) [reply .github/workflows/docs.yml]: Mm, I see. Yeah, that's alright for now I think.

Here's an idea: we could use something like `ansi2txt` from https://github.com/kilobyte/colorized-logs to take the colored output and remove it for the markdown report.

(I can't see where that feature is btw, might be missing something).

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (jedevc) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: > This is possible (with some implementation work) BUT it will NOT change the number of machines started, because each job gets its own machine, regardless of workflow grouping.

Yeah, fair enough - I guess it's more just for sorting at the github level. I don't *currently* have the time to PR this in (reviews + flakes + prep for v0.13 are most of my attention this week) - but can try and find a moment after that's landed if no one else can pick it up.

> There is no benefit to this. It should be done at the Dagger level.

Fair. I think I agree - this does mean that if we want lots of tasks done, you need to use your language primitives for it. This is good.

But, I think if we're doing this more, the performance of the traces view in dagger cloud might need some help - when you have a *lot* of spans together, the performance can really suffer, and the page takes way more time to load than the basic github actions logs view. But hopefully we can feel the pain of this and do something here.

File: .github/dagger.json
Comment (jedevc) [inline .github/dagger.json:7]: What's the plan for where this lives long-term btw? Could we move it to the `dagger` org or under `modules/`? Not a blocker at all, just curious, personally, I prefer having it "semi-official", with permissions/reviews managed by the same system we use would be good.

File: .github/main.go
Comment (matipan) [inline .github/main.go:38]: Thinking a bit how this can be translated to the idea of self contained functions in the context of pocketci.

Each of the pipelines described here can be treated as its own self contained function that defines when it should be triggered and where it should run. At the moment we are separating this even further (although I know you already think its overkill) since each step of SDK jobs (lint/test/publish) is its own pipeline with a separate dagger call. If we could bring everything into one we could take an approach that would yield numerous dagger functions (this is still experimental):
```go
// Function only gets called when there is a pull request (opened,reopened,synchronize,ready_for_review) or a push event (to branch main).
func (dev *DaggerDev) SdkGoOnGithub(ctx context.Context,
	src *dagger.Directory,
	// +default="pull_request,push"
	event string,
	// +default="opened,reopened,synchronize,ready_for_review,main"
	filter string,
	eventTrigger *dagger.File) error {
	sdk := dev.SDK().Go

	if err := sdk.Lint(ctx); err != nil {
		return err
	}

	return sdk.Test(ctx)
}
```

Each SDK would have its own function with virtually with the same trigger. This is definitely more verbose than what we currently have (and we still have not solved the compute problem). I'm also not liking so much how the filtering of events and actions within an event is being built, there is some more thinking I need to do there. Its worth doing a bit more thinking and playing with dagger/dagger more to see how quickly the idea of functions defining its own trigger logic breaks.

File: .github/workflows/docs.yml
Comment (jedevc) [reply .github/workflows/docs.yml]: Ah, I found it. FYI, there's still some control color codes in there sometimes: https://github.com/dagger/dagger/actions/runs/10648614702?pr=8241

File: .github/dagger.json
Comment (shykes) [reply .github/dagger.json:7]: Yes, we should move it in. There are a few options, I will open a follow-up proposal.

File: .github/main.go
Comment (shykes) [reply .github/main.go:38]: This won't work IMO. You need to keep the CI-specific logic (event routing etc) decoupled, at a higher layer.

File: .github/workflows/docs.yml
Comment (shykes) [reply .github/workflows/docs.yml]: Could it be because of nested dagger engines?

**Comment by shykes** [comment]
@matipan @gerhard @jedevc I have `elixir test` and `engine test specific` failing, are those flakes?

**Comment by jedevc** [comment]
Hm, no the failing elixir job looks genuine, never seen this failure before. cc @wingyplus who might have some ideas.

It looks like:
- We see a bunch of `ERR failed to emit telemetry error="traces export: failed to send to http://127.0.0.1:39707/v1/traces: 500 Internal Server Error`
- And also get a test failure?

`engine test specific` failing looks like a known flake to me.

---

One unrelated note: the new checks names have duplication in them :thinking: 

image

Ideally the job grouping would remedy some of that, but in the meantime, could we have something like `elixir sdk: test / check` with `check` (or similar) being the job name?

**Comment by wingyplus** [comment]
> Hm, no the failing elixir job looks genuine, never seen this failure before. cc @wingyplus who might have some ideas.
> 
> It looks like:
> 
> * We see a bunch of `ERR failed to emit telemetry error="traces export: failed to send to http://127.0.0.1:39707/v1/traces: 500 Internal Server Error`
> * And also get a test failure?
> 
> `engine test specific` failing looks like a known flake to me.
> 
> One unrelated (non-blocker) note: the new checks names have duplication in them 🤔
> 
> image
> 
> Ideally the job grouping would remedy some of that, but in the meantime, could we have something like `elixir sdk: test / check` with `check` (or similar) being the job name?

Will looking into that problem in around 3-4 hours. ✌️

**Comment by shykes** [comment]
> Ideally the job grouping would remedy some of that, but in the meantime, could we have something like `elixir sdk: test / check` with `check` (or similar) being the job name?

Sadly we cannot. Initially the naming was exactly what you suggested. But the job name is used to match "required checks". The Cubzh team use this feature and after much debugging, we landed on this.

We *could* make that behavior configurable with a flag, although it would be an obscure one.

Can we make this a followup? 🙏

**Comment by wingyplus** [comment]
@jedevc From what I can tell. The test failed on [mounted cache test]( https://github.com/dagger/dagger/blob/main/sdk/elixir/test/dagger/client_test.exs#L125). And from this [failure](https://github.com/dagger/dagger/actions/runs/10648614702/job/29517873407#step:5:3141), it looks like cache volume is stored from the previous run somehow cause it gets double result. I open a PR to fix this at https://github.com/dagger/dagger/pull/8317 by using LOCKED mode but I'm also not sure if it's the right fix.

**Comment by shykes** [comment]
> @jedevc From what I can tell. The test failed on [mounted cache test](https://github.com/dagger/dagger/blob/main/sdk/elixir/test/dagger/client_test.exs#L125). And from this [failure](https://github.com/dagger/dagger/actions/runs/10648614702/job/29517873407#step:5:3141), it looks like cache volume is stored from the previous run somehow cause it gets double result. I open a PR to fix this at #8317 by using LOCKED mode but I'm also not sure if it's the right fix.

Is it possible that my PR caused the issue? Maybe a subtle change in the YAML config?

File: .github/workflows/elixir-sdk-lint-dev-engine.yml
Comment (shykes) [reply .github/workflows/elixir-sdk-lint-dev-engine.yml:1]: > But, I think if we're doing this more, the performance of the traces view in dagger cloud might need some help - when you have a lot of spans together, the performance can really suffer, and the page takes way more time to load than the basic github actions logs view. But hopefully we can feel the pain of this and do something here.

I believe @vito has been focusing specifically on Traces UI performance :)

File: .github/main.go
Comment (shykes) [reply .github/main.go:61]: @jedevc this is what I initially want to do, it most closely models the reality of the underlying data structure. But, it wouldn't be usable from the CLI. I am setting a constraint for myself to make the API great in both code and the CLI.

File: dagger.json
Comment (wingyplus) [inline dagger.json:51]: Why this version got revert back to 0.12.6?

File: .github/main.go
Comment (jedevc) [reply .github/main.go:61]: tangent: is this then an argument to make the CLI more powerful, so it can handle construction of more complex objects, i.e. ones that take other objects as parameters?

cc @helderco

File: .dagger/main.go
Comment (jedevc) [inline .dagger/main.go:106]: :eyes: just wanted to check, is this being brought back before/after we merge this? This does catch real issues, so don't want to leave it too long.

Comment (shykes) [reply .dagger/main.go:106]: Sorry yes, that is WIP, will bring back today

File: dagger.json
Comment (shykes) [reply dagger.json:51]: (no idea how this happened btw)

File: .github/main.go
Comment (shykes) [reply .github/main.go:61]: > tangent: is this then an argument to make the CLI more powerful, so it can handle construction of more complex objects, i.e. ones that take other objects as parameters?

Yes :)

File: .dagger/main.go
Comment (shykes) [reply .dagger/main.go]: Done.

**Comment by shykes** [comment]
@gerhard in an effort to make you happy, we are now down to 3k lines of YAML instead of 11k :)

**Comment by shykes** [comment]
Force-pushing a small change to re-run all failed workflows...

**Comment by shykes** [comment]
FYI I am combining #8284 into this, it's too much work to maintain both in parallel.

File: .github/workflows/elixir-dev.gen.yml
Comment (wingyplus) [inline .github/workflows/elixir-dev.gen.yml:15]: The `github.workflow` is a workflow name. So then how is it important to keep this because generated workflow has only single job?

Comment (shykes) [reply .github/workflows/elixir-dev.gen.yml:15]: It's to control concurrency of multiple instances of the workflow. Typically if you push 10 times to the same PR, that triggers 10 different instances of the worfklow. With the concurrency group, only 1 instance will run at a time.

Comment (wingyplus) [reply .github/workflows/elixir-dev.gen.yml:15]: Ahh, I see. Thanks for the info.

**Pull request merged**
